### PR TITLE
Media binary

### PR DESCRIPTION
**Pull request by srascar**
Add route in api to update a media binary content

### CODE REVIEW

File: Provider/FileProvider.php [hunk 1]
```diff
@@ -382,15 +403,67 @@ public function validate(ErrorElement $errorElement, MediaInterface $media)
...
+            throw new \RuntimeException('Expected Request in binary content');
+        }
+
+        $content = $request->getContent();
+        // create unique id for media reference
+        $guesser = ExtensionGuesser::getInstance();
+        $extension = $guesser->guess($media->getContentType());
+        $reference = $this->generateMediaUniqId($media).'.'.$extension;
+
+        $tmpFile = fopen(sys_get_temp_dir().DIRECTORY_SEPARATOR.$reference, 'w+');
```
Comment (rande) [inline Provider/FileProvider.php]: You don't close the file pointer and the file will not be removed....

Comment (srascar) [inline Provider/FileProvider.php]: files are removed in the post persist/update functions

File: Provider/FileProvider.php [hunk 2]
```diff
@@ -382,15 +403,68 @@ public function validate(ErrorElement $errorElement, MediaInterface $media)
...
+        $metaData = stream_get_meta_data($tmpFile);
+        fwrite($tmpFile, $content);
+        fclose($tmpFile);
+        $file =  new SonataMediaFile($metaData['uri']);
```
Comment (rande) [inline Provider/FileProvider.php]: Can you name it ``ApiMediaFile`` ? Thanks

**Comment by srascar** [comment]
@rande modifications ready

File: Provider/FileProvider.php [hunk 3]
```diff
@@ -169,27 +173,27 @@ public function postUpdate(MediaInterface $media)
...
-                throw new \RuntimeException('The file does not exist : ' . $media->getBinaryContent());
-            }
+        if ($media->getBinaryContent() instanceof Request) {
+            $this->generateBinaryFromRequest($media);
+
+            return;
+        }
 
+        // if the binary content is a filename => convert to a valid File
+        if (is_file($media->getBinaryContent())) {
```
Comment (Bladrak) [inline Provider/FileProvider.php]: You still need to throw the exception if the binary content is not a file afterwards, don't you?

**Comment by Bladrak** [comment]
Is there a sandbox PR with the functional tests?

File: Provider/FileProvider.php (hunk 3 above)
Comment (srascar) [inline Provider/FileProvider.php]: done

Comment (rande) [inline Provider/FileProvider.php]: it is better to reverse the if:
```
if(!is_file(...)) {
  throw new Exception ...
}

**Comment by srascar** [comment]
@rande  done

File: Controller/Api/MediaController.php [hunk 4]
```diff
@@ -366,4 +396,26 @@ protected function handleWriteMedium($request, MediaInterface $medium, MediaProv
...
+    protected function handleRequestBinaryContent(MediaInterface $media, Request $request)
+    {
+        try {
+            $mediaProvider = $this->mediaPool->getProvider($media->getProviderName());
+        } catch (\RuntimeException $ex) {
+            throw new NotFoundHttpException($ex->getMessage(), $ex);
+        }
+
+        $media->setBinaryContent($request);
+        $mediaProvider->transform($media);
```
Comment (rande) [inline Controller/Api/MediaController.php]: Do we need this transform here ?

**Comment by srascar** [comment]
@rande PR updated

File: Provider/FileProvider.php [hunk 5]
```diff
@@ -382,15 +396,58 @@ public function validate(ErrorElement $errorElement, MediaInterface $media)
...
+        $guesser = ExtensionGuesser::getInstance();
+        $extension = $guesser->guess($media->getContentType());
+
+        if (!$extension) {
+            throw new \RuntimeException(
+                sprintf('Unable to guess extension for content type %s', $media->getContentType())
+            );
+        }
+
+        if (php_sapi_name() === 'cli') {
```
Comment (rande) [inline Provider/FileProvider.php]: This should be done at the top of this function

File: Extra/ApiMediaFile.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\MediaBundle\Extra;
+
+
+use Symfony\Component\HttpFoundation\File\File;
+
+class ApiMediaFile extends File
```
Comment (rande) [inline Extra/ApiMediaFile.php]: This file is not complete

Comment (srascar) [inline Extra/ApiMediaFile.php]: fixed

**Pull request merged**